File: android/app/src/main/java/com/zulipmobile/notifications/NotifyReact.java
package com.zulipmobile.notifications;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.Log;
import com.facebook.react.ReactApplication;
import com.facebook.react.ReactNativeHost;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.common.LifecycleState;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import com.zulipmobile.MainActivity;

import static com.zulipmobile.notifications.NotificationHelper.TAG;

/**
 * Methods for telling React about a notification.
 *
 * This logic was largely inherited from the wix library.
 * TODO: Replace this with a fresh implementation based on RN upstream docs.
 */
class NotifyReact {

    static void notifyReact(ReactApplication application, final Bundle data) {
        NotificationsModule.initialNotification = data;

        if (!emitIfResumed(application, "notificationOpened", Arguments.fromBundle(data))) {
            // The app will check initialNotification on launch.
            Log.d(TAG, "notifyReact: launching main activity");
            launchMainActivity((Context) application);
        }
    }

    private static boolean emitIfResumed(ReactApplication application, final String eventName, final @Nullable Object data) {
        final ReactNativeHost host = application.getReactNativeHost();
        if (!host.hasInstance()) {
            // Calling getReactInstanceManager would try to create one...
            // which asserts we're on the UI thread, which isn't true if we
            // got here from a Service.
            return false;
        }
        final ReactContext reactContext = host.getReactInstanceManager().getCurrentReactContext();
        if (reactContext == null
                || !reactContext.hasActiveCatalystInstance()
                || reactContext.getLifecycleState() != LifecycleState.RESUMED) {
            return false;
        }
        emit(reactContext, eventName, data);
        return true;
    }

    static void emit(@NonNull ReactContext reactContext, String eventName, @Nullable Object data) {
        reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(eventName, data);
    }

    private static void launchMainActivity(Context context) {
        final Intent intent = new Intent(context, MainActivity.class);
        // See these sections in the Android docs:
        //   https://developer.android.com/guide/components/activities/tasks-and-back-stack#TaskLaunchModes
        //   https://developer.android.com/reference/android/content/Intent#FLAG_ACTIVITY_CLEAR_TOP
        //
        // * The flag FLAG_ACTIVITY_NEW_TASK is redundant in that it produces the
        //   same effect as setting `android:launchMode="singleTask"` on the
        //   activity, which we've done; but Context#startActivity requires it for
        //   clarity's sake, a requirement overridden in Activity#startActivity,
        //   because the behavior without it only makes sense when starting from
        //   an Activity.  Our `context` is a service, so it's required.
        //
        // * The flag FLAG_ACTIVITY_CLEAR_TOP is mentioned as being what the
        //   notification manager does; so use that.  It has no effect as long
        //   as we only have one activity; but if we add more, it will destroy
        //   all the activities on top of the target one.
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK
                | Intent.FLAG_ACTIVITY_CLEAR_TOP);
        context.startActivity(intent);
    }
}

File: android/app/src/main/java/com/zulipmobile/notifications/NotifyReact.kt
@file:JvmName("NotifyReact")

package com.zulipmobile.notifications

import android.content.Intent
import android.os.Bundle
import android.util.Log
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReactContext
import com.zulipmobile.*

/**
 * Methods for telling React about a notification.
 *
 * This logic was largely inherited from the wix library.
 * TODO: Replace this with a fresh implementation based on RN upstream docs.
 */

/**
 * Recognize if an ACTION_VIEW intent came from tapping a notification; handle it if so
 *
 * Just if the intent is recognized,
 * sends a 'notificationOpened' event to the JavaScript layer
 * and returns true.
 * Else does nothing and returns false.
 *
 * Do not call if `intent.action` is not ACTION_VIEW.
 */
internal fun maybeHandleViewNotif(intent: Intent, maybeReactContext: ReactContext?): Boolean {
    assert(intent.action == Intent.ACTION_VIEW)

    val url = intent.data
    // Launch MainActivity on tapping a notification
    if (url?.scheme == "zulip" && url.authority == NOTIFICATION_URL_AUTHORITY) {
        val data = intent.getBundleExtra(EXTRA_NOTIFICATION_DATA) ?: return false
        logNotificationData("notif opened", data)
        notifyReact(maybeReactContext, data)
        return true
    }

    return false
}

internal fun notifyReact(reactContext: ReactContext?, data: Bundle) {
    // TODO deduplicate this with handleSend in SharingHelper.kt; see
    // https://github.com/zulip/zulip-mobile/pull/5146#discussion_r764437055
    // Until then, keep in sync when changing.
    val appStatus = reactContext?.appStatus
    Log.d(TAG, "notifyReact: app status is $appStatus")
    when (appStatus) {
        null, ReactAppStatus.NOT_RUNNING ->
            // Either there's no JS environment running, or we haven't yet reached
            // foreground.  Expect the app to check initialNotification on launch.
            NotificationsModule.initialNotification = data
        ReactAppStatus.BACKGROUND, ReactAppStatus.FOREGROUND ->
            // JS is running and has already reached foreground.  It won't check
            // initialNotification again, but it will see a notificationOpened event.
            reactContext.emitEvent("notificationOpened", Arguments.fromBundle(data))
    }
}
